rpc: add minconf/maxconf options to sendall and fund transaction calls

ishaanam opened the pull request:
This PR adds a "minconf" option to `fundrawtransaction`, `walletcreatefundedpsbt`,  and `sendall`.
Alternative implementation of #14641
Fixes #14542

Edit: This PR now also adds this option to `send`

champo committed:
Wallet/RPC: Allow specifying min & max chain depth for inputs used by fund calls

Enables users to craft BIP-125 replacements with changes to the output
list, ensuring that if additional funds are needed they will be added.

1440000bytes reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/25375/commits/0f023da016b2230af2d86b9461bfe5218716fdb5

Tested `minconf` argument on regtest with:

```
$ bitcoin-cli generatetoaddress 102 bcrt1qjjdvpcz3l2p4q2y7zc6ztkk3rvagke248y6ckm

$ bitcoin-cli listunspent
[
  {
    "txid": "56a91239cfcff53da811ae1969bb6e8e288f15ab58cac8fe455ea906a3b51ecb",
    "vout": 0,
    "address": "bcrt1qjjdvpcz3l2p4q2y7zc6ztkk3rvagke248y6ckm",
    "label": "",
    "scriptPubKey": "0014949ac0e051fa8350289e163425dad11b3a8b6555",
    "amount": 50.00000000,
    "confirmations": 101,
    "spendable": true,
    "solvable": true,
    "desc": "wpkh([ae30e814/84'/1'/0'/0/0]020a3db12de02b1f4bbb5e5016f7cebaa427e0c60d20b95ebc552ceb97c86f3722)#n2vukszk",
    "safe": true
  },
  {
    "txid": "f1257f379856d91401d27229dbdf797802a521edb6f958bc884e316265a48cfc",
    "vout": 0,
    "address": "bcrt1qjjdvpcz3l2p4q2y7zc6ztkk3rvagke248y6ckm",
    "label": "",
    "scriptPubKey": "0014949ac0e051fa8350289e163425dad11b3a8b6555",
    "amount": 50.00000000,
    "confirmations": 102,
    "spendable": true,
    "solvable": true,
    "desc": "wpkh([ae30e814/84'/1'/0'/0/0]020a3db12de02b1f4bbb5e5016f7cebaa427e0c60d20b95ebc552ceb97c86f3722)#n2vukszk",
    "safe": true
  }
]

```

```
$ bitcoin-cli createrawtransaction '[]' '{"data":"505220233235333735"}'

02000000000100000000000000000b6a0950522023323533373500000000
```
```


$ bitcoin-cli fundrawtransaction "02000000000100000000000000000b6a0950522023323533373500000000" "{\"minconf\": 102}"
{
  "hex": "0200000001fc8ca46562314e88bc58f9b6ed21a5027879dfdb2972d20114d95698377f25f10000000000feffffff0272f1052a010000002251209b08765d5b558137195f9b3b5b4f5ceaffddcac91590e76f904fdde08cc033bb00000000000000000b6a0950522023323533373500000000",
  "fee": 0.00000142,
  "changepos": 0
}

$ bitcoin-cli decoderawtransaction 0200000001fc8ca46562314e88bc58f9b6ed21a5027879dfdb2972d20114d95698377f25f10000000000feffffff0272f1052a010000002251209b08765d5b558137195f9b3b5b4f5ceaffddcac91590e76f904fdde08cc033bb00000000000000000b6a0950522023323533373500000000
{
  "txid": "4974082891c14f96170f2af7f381e3a6f9510cb9e42841b1e12463f09b164ad5",
  "hash": "4974082891c14f96170f2af7f381e3a6f9510cb9e42841b1e12463f09b164ad5",
  "version": 2,
  "size": 114,
  "vsize": 114,
  "weight": 456,
  "locktime": 0,
  "vin": [
    {
      "txid": "f1257f379856d91401d27229dbdf797802a521edb6f958bc884e316265a48cfc",
      "vout": 0,
      "scriptSig": {
        "asm": "",
        "hex": ""
      },
      "sequence": 4294967294
    }
  ],
  "vout": [
    {
      "value": 49.99999858,
      "n": 0,
      "scriptPubKey": {
        "asm": "1 9b08765d5b558137195f9b3b5b4f5ceaffddcac91590e76f904fdde08cc033bb",
        "desc": "addr(bcrt1pnvy8vh2m2kqnwx2lnva4kn6uatlamjkfzkgwwmusflw7prxqxwas8y08ed)#4mpq27kr",
        "hex": "51209b08765d5b558137195f9b3b5b4f5ceaffddcac91590e76f904fdde08cc033bb",
        "address": "bcrt1pnvy8vh2m2kqnwx2lnva4kn6uatlamjkfzkgwwmusflw7prxqxwas8y08ed",
        "type": "witness_v1_taproot"
      }
    },
    {
      "value": 0.00000000,
      "n": 1,
      "scriptPubKey": {
        "asm": "OP_RETURN 505220233235333735",
        "desc": "raw(6a09505220233235333735)#wxsnvgu2",
        "hex": "6a09505220233235333735",
        "type": "nulldata"
      }
    }
  ]
}
```

```
$ bitcoin-cli fundrawtransaction "02000000000100000000000000000b6a0950522023323533373500000000" "{\"minconf\": 103}"
error code: -4
error message:
Insufficient funds
```
nit: If no UTXO available in wallet with `minconf` confirmations, the error could be specific to help the user. Example: Insufficient funds with required confirmations

achow101 commented:
> nit: If no UTXO available in wallet with `minconf` confirmations, the error could be specific to help the user. Example: Insufficient funds with required confirmations

It's actually quite difficult currently to get more specific errors out of coin selection. There's quite a bit of plumbing needed that does not currently exist. #25218 and #24845 add a lot of what is needed to get such errors, but there's still an additional issue where it is hard to tell whether the insufficient funds is due to this particular filter or something else in coin selection.

w0xlt reviewed (approved):
Code Review ACK https://github.com/bitcoin/bitcoin/pull/25375/commits/0f023da016b2230af2d86b9461bfe5218716fdb5

w0xlt commented:
Perhaps this same option could be included in `send()` RPC ? 
This also creates a `CCoinControl` object and passes it to `FundTransaction()`, just like the other methods covered in this PR.

luke-jr commented:
This seems like a subset of #22049

ishaanam commented:
Changes made since last push:
- refactored sendall tests based on feedback from @Xekyo 
- moved minconf help text to FundTxDoc
- added minconf option to `send` as suggested by @w0xlt and added a test for that
- added check that minconf is not negative and added tests to ensure that an error is raised when minconf is negative

ishaanam commented:
I'm closing this PR as it is quite similar to #22049

luke-jr commented:
@ishaanam As #22049 has been abandoned, perhaps copy and rebase it here?

ishaanam committed:
rpc: add minconf and maxconf options to sendall

1440000bytes commented:
> @ishaanam As #22049 has been abandoned, perhaps copy and rebase it here?
> 
> P.S. I still have a rebased version at [master...luke-jr:rpc_fundtx_minmaxconf](https://github.com/bitcoin/bitcoin/compare/master...luke-jr:rpc_fundtx_minmaxconf)

I agree with @luke-jr , its important to fix the issue not the author or pr

ishaanam commented:
> @ishaanam As #22049 has been abandoned, perhaps copy and rebase it here?
> 
> P.S. I still have a rebased version at [master...luke-jr:rpc_fundtx_minmaxconf](https://github.com/bitcoin/bitcoin/compare/master...luke-jr:rpc_fundtx_minmaxconf)

I have reopened this PR, thanks for rebasing #22049!

1440000bytes reviewed (commented):
reACK https://github.com/bitcoin/bitcoin/pull/25375/commits/3f0f67c1f478584a8225b8e02ba75d3920634200

ishaanam commented:
Rebased

1440000bytes reviewed (approved):
reACK https://github.com/bitcoin/bitcoin/pull/25375/commits/a7ab0960ab92171cae72e9371b8d3e63a2f5c41a

furszy commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -1338,6 +1339,68 @@ def test_feerate_rounding(self):
+
+        self.log.info("Fail to craft a new TX that sends more funds with add_inputs = False")
+        raw_tx2 = wallet.createrawtransaction([{'txid': utxo1['txid'], 'vout': utxo1['vout']}], {target_address: 1})
+        assert_raises_rpc_error(-4, "Insufficient funds", wallet.fundrawtransaction, raw_tx2, {'fee_rate': 1, 'add_inputs': False})
```
Any reason for adding this case here?

Seems to not be needed as it's already covered inside the `test_add_inputs_default_value` function and we here want to test the min/max conf params.

ishaanam replied:
This commit was authored by @champo before `test_add_inputs_default_value` was added. I have removed this test.

furszy reviewed (commented):
Concept ACK

Left few nits, nothing important. Will do a deeper round in the coming days.

furszy commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1300,6 +1300,8 @@ RPCHelpMan sendall()
                         {"lock_unspents", RPCArg::Type::BOOL, RPCArg::Default{false}, "Lock selected unspent outputs"},
                         {"psbt", RPCArg::Type::BOOL,  RPCArg::DefaultHint{"automatic"}, "Always return a PSBT, implies add_to_wallet=false."},
                         {"send_max", RPCArg::Type::BOOL, RPCArg::Default{false}, "When true, only use UTXOs that can pay for their own fees to maximize the output amount. When 'false' (default), no UTXO is left behind. send_max is incompatible with providing specific inputs."},
+                        {"minconf", RPCArg::Type::NUM, RPCArg::Default{1}, "If add_inputs is specified, require inputs with at least this confirmations."},
```
I maybe missed it but where the `minconf` is set to 1 inside `sendall`? (default value is 0)

ishaanam replied:
The default `minconf` in `sendall` is 1 because `sendall` does not spend unconfirmed inputs.

furszy replied:
where is that established?

let's say that the user does not pass `minconf`, so the default `min_conf` value inside coin control is 0. So, the `AvailableCoins` call will return outputs with depth 0 as well, which are added to the raw tx.

ishaanam replied:
`AvailableCoins` won't return unconfirmed utxos that aren't change unless `include_unsafe` is set to true. However, because this means that `sendall` can still spend unconfirmed change, which means that the default `minconf` is really 0. I have the docs to reflect this.

brunoerg commented on test/functional/rpc_psbt.py:
```diff
@@ -105,6 +106,70 @@ def test_utxo_conversion(self):
+        assert_equal(len(tx1_inputs), 1)
+
+        utxo1 = tx1_inputs[0]
+        assert unconfirmed_txid == utxo1['txid']
```
```suggestion
        assert_equal(unconfirmed_txid, utxo1['txid'])
```

ishaanam replied:
Done

brunoerg commented on test/functional/rpc_psbt.py:
```diff
@@ -105,6 +106,70 @@ def test_utxo_conversion(self):
+        wallet = self.nodes[0].get_wallet_rpc("minconf")
+
+        # Fund the wallet with different chain heights
+        self.nodes[1].sendtoaddress(wallet.getnewaddress(), 1)
```
maybe you could use a for loop here to avoid repeat same code?

ishaanam commented on test/functional/rpc_psbt.py:
```diff
@@ -105,6 +106,70 @@ def test_utxo_conversion(self):
+        self.generate(self.nodes[1], 1)
+        self.nodes[1].sendtoaddress(wallet.getnewaddress(), 1)
+        self.nodes[1].sendtoaddress(wallet.getnewaddress(), 1)
+        self.generate(self.nodes[1], 1)
```
Done

ishaanam commented:
@furszy, @brunoerg, Thanks for the review, I have implemented all of your suggestions.

furszy commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -1338,6 +1339,62 @@ def test_feerate_rounding(self):
+        wallet = self.nodes[0].get_wallet_rpc("minconf")
+
+        # Fund the wallet with different chain heights
+        for _ in range(2):
+            self.nodes[2].sendmany("", {wallet.getnewaddress():1, wallet.getnewaddress():1})
+            self.generate(self.nodes[2], 1)
+
+        self.sync_blocks()
```
in 0e1f99c4:

nit: can remove the `sync_blocks` call. `self.generate` calls `sync_all()` internally if no sync function is provided.

same in the other test

ishaanam replied:
Done

furszy reviewed (commented):
The RPC API changes will need release-notes too.

ishaanam commented:
Added release notes.

1440000bytes reviewed (approved):
reACK https://github.com/bitcoin/bitcoin/pull/25375/commits/4d7a1b6d7245c24a0749b04a975bcd4f630bb304

brunoerg commented on doc/release-notes-25375.md:
```diff
@@ -0,0 +1,11 @@
+Updated RPCs
+--------
+
+The `minconf`option, which allows a user to specify the minimum number
```
nit:
```suggestion
The `minconf` option, which allows a user to specify the minimum number
```

ishaanam replied:
Fixed

ishaanam commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -1338,6 +1339,60 @@ def test_feerate_rounding(self):
+        self.log.info("Fail to broadcast a new TX with maxconf 0 due to BIP125 rules to verify it actually chose unconfirmed outputs")
+        funded_invalid = wallet.fundrawtransaction(raw_tx2, {'add_inputs': True, 'maxconf': 0, 'fee_rate': 10})['hex']
+        final_invalid = wallet.signrawtransactionwithwallet(funded_invalid)['hex']
+        assert_raises_rpc_error(-26, "bad-txns-spends-conflicting-tx", self.nodes[0].sendrawtransaction, final_invalid)
```
I have added this comment.

furszy commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1300,6 +1300,8 @@ RPCHelpMan sendall()
                         {"lock_unspents", RPCArg::Type::BOOL, RPCArg::Default{false}, "Lock selected unspent outputs"},
                         {"psbt", RPCArg::Type::BOOL,  RPCArg::DefaultHint{"automatic"}, "Always return a PSBT, implies add_to_wallet=false."},
                         {"send_max", RPCArg::Type::BOOL, RPCArg::Default{false}, "When true, only use UTXOs that can pay for their own fees to maximize the output amount. When 'false' (default), no UTXO is left behind. send_max is incompatible with providing specific inputs."},
+                        {"minconf", RPCArg::Type::NUM, RPCArg::Default{0}, "If add_inputs is specified, require inputs with at least this confirmations."},
+                        {"maxconf", RPCArg::Type::NUM, RPCArg::Optional::OMITTED, "If add_inputs is specified, require inputs with at most this confirmations."},
```
In b3a3cb14:

`sendall` has no `add_inputs` argument. If `inputs` are specified, the command will only select them.

ishaanam replied:
Good catch, I have changed the docs to reflect this.

furszy reviewed (commented):
Almost there, code reviewed b3a3cb14.
Left a last finding and it's ready to go.

furszy reviewed (approved):
ACK 24c906b

achow101 commented:
ACK 24c906b3311336ccaf5ea9c70da5ebde6f1cd041

Xekyo commented:
I noticed that https://github.com/bitcoin/bitcoin/commit/1e14aeacd3ae86deb23ad4255cafbd0a3d9a894a [added a test for `send` in test/functional/wallet_send.py](https://github.com/bitcoin/bitcoin/compare/0f023da016b2230af2d86b9461bfe5218716fdb5..1e14aeacd3ae86deb23ad4255cafbd0a3d9a894a#diff-45c9fdef949aa07c272e8b6b7fb6a6d7db3ba056ba5746c6feba6ba90a58b85d), but that test does not seem to be part of the latest change set. Is it possible that the send test got lost accidentally or did I miss the reasoning why it got removed?

murchandamus reviewed (commented):
I noticed that https://github.com/bitcoin/bitcoin/commit/1e14aeacd3ae86deb23ad4255cafbd0a3d9a894a [added a test for `send` in test/functional/wallet_send.py](https://github.com/bitcoin/bitcoin/compare/0f023da016b2230af2d86b9461bfe5218716fdb5..1e14aeacd3ae86deb23ad4255cafbd0a3d9a894a#diff-45c9fdef949aa07c272e8b6b7fb6a6d7db3ba056ba5746c6feba6ba90a58b85d), but that test does not seem to be part of the latest change set. Is it possible that the send test got lost accidentally or did I miss the reasoning why it got removed?

ishaanam commented:
@Xekyo good catch. I replaced some of my commits with those from #22049 and it looks like that didn't include a test for 'send'. I'll add that back soon.

Xekyo commented on src/wallet/rpc/spend.cpp:
```diff
@@ -744,6 +762,8 @@ RPCHelpMan fundrawtransaction()
                             {"include_unsafe", RPCArg::Type::BOOL, RPCArg::Default{false}, "Include inputs that are not safe to spend (unconfirmed transactions from outside keys and unconfirmed replacement transactions).\n"
                                                           "Warning: the resulting transaction may become invalid if one of the unsafe inputs disappears.\n"
                                                           "If that happens, you will need to fund the transaction with different inputs and republish it."},
+                            {"minconf", RPCArg::Type::NUM, RPCArg::Default{0}, "If add_inputs is specified, require inputs with at least this confirmations."},
+                            {"maxconf", RPCArg::Type::NUM, RPCArg::Optional::OMITTED, "If add_inputs is specified, require inputs with at most this confirmations."},
```
Nit if you need to update anything else. 
Here and in the other manual entries, this reads a bit odd to me. Maybe:

```suggestion
                            {"minconf", RPCArg::Type::NUM, RPCArg::Default{0}, "If add_inputs is specified, require inputs with at least this many confirmations."},
                            {"maxconf", RPCArg::Type::NUM, RPCArg::Optional::OMITTED, "If add_inputs is specified, require inputs with at most this many confirmations."},
```

ishaanam replied:
I've changed all of them to this.

Xekyo commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1140,6 +1160,8 @@ RPCHelpMan send()
                     {"include_unsafe", RPCArg::Type::BOOL, RPCArg::Default{false}, "Include inputs that are not safe to spend (unconfirmed transactions from outside keys and unconfirmed replacement transactions).\n"
                                                           "Warning: the resulting transaction may become invalid if one of the unsafe inputs disappears.\n"
                                                           "If that happens, you will need to fund the transaction with different inputs and republish it."},
+                    {"minconf", RPCArg::Type::NUM, RPCArg::Default{0}, "If add_inputs is specified, require inputs with at least this confirmations."},
+                    {"maxconf", RPCArg::Type::NUM, RPCArg::Optional::OMITTED, "If add_inputs is specified, require inputs with at most this confirmations."},
```
Nit as above:

```suggestion
                    {"minconf", RPCArg::Type::NUM, RPCArg::Default{0}, "If add_inputs is specified, require inputs with at least this many confirmations."},
                    {"maxconf", RPCArg::Type::NUM, RPCArg::Optional::OMITTED, "If add_inputs is specified, require inputs with at most this many confirmations."},
```

ishaanam replied:
Done

Xekyo commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1596,6 +1639,8 @@ RPCHelpMan walletcreatefundedpsbt()
                             {"include_unsafe", RPCArg::Type::BOOL, RPCArg::Default{false}, "Include inputs that are not safe to spend (unconfirmed transactions from outside keys and unconfirmed replacement transactions).\n"
                                                           "Warning: the resulting transaction may become invalid if one of the unsafe inputs disappears.\n"
                                                           "If that happens, you will need to fund the transaction with different inputs and republish it."},
+                            {"minconf", RPCArg::Type::NUM, RPCArg::Default{0}, "If add_inputs is specified, require inputs with at least this confirmations."},
+                            {"maxconf", RPCArg::Type::NUM, RPCArg::Optional::OMITTED, "If add_inputs is specified, require inputs with at most this confirmations."},
```
As above:

```suggestion
                            {"minconf", RPCArg::Type::NUM, RPCArg::Default{0}, "If add_inputs is specified, require inputs with at least this many confirmations."},
                            {"maxconf", RPCArg::Type::NUM, RPCArg::Optional::OMITTED, "If add_inputs is specified, require inputs with at most this many confirmations."},
```

ishaanam replied:
Done

murchandamus reviewed (commented):
ACK 41ebc2b0294941d12980e19307bd991f9ff85301

murchandamus reviewed (approved):
ACK cfe5aebc79c510bd2156e199c3324d7ee1f8d2ad

achow101 commented:
ACK cfe5aebc79c510bd2156e199c3324d7ee1f8d2ad

furszy commented on test/functional/wallet_send.py:
```diff
@@ -487,6 +489,16 @@ def run_test(self):
 
+        self.log.info("Minconf")
+        self.nodes[1].createwallet(wallet_name="minconfw")
+        minconfw= self.nodes[1].get_wallet_rpc("minconfw")
```
nit:
missing space
```suggestion
        minconfw = self.nodes[1].get_wallet_rpc("minconfw")
```

furszy reviewed (approved):
diff ACK cfe5aebc, only a non-blocking nit.

MarcoFalke commented on src/wallet/rpc/spend.cpp:
```diff
@@ -744,6 +762,8 @@ RPCHelpMan fundrawtransaction()
                                                           "Warning: the resulting transaction may become invalid if one of the unsafe inputs disappears.\n"
                                                           "If that happens, you will need to fund the transaction with different inputs and republish it."},
+                            {"minconf", RPCArg::Type::NUM, RPCArg::Default{0}, "If add_inputs is specified, require inputs with at least this many confirmations."},
+                            {"maxconf", RPCArg::Type::NUM, RPCArg::Optional::OMITTED, "If add_inputs is specified, require inputs with at most this many confirmations."},
```
nit: Looks like this should be documented as (optional) in the RPC doc output. In theory this instance can be fixed by replacing OMITTED with OMITTED_NAMED_ARG, however a more complete fix is in https://github.com/bitcoin/bitcoin/pull/26706

(Same for other `"maxconf"` in this pull)

harding commented:
What was the motivation for the `maxconf` parameter?  I'm guessing it's was the result of @achow101's [comment](https://github.com/bitcoin/bitcoin/pull/22049#pullrequestreview-667127883), but I'm having trouble imagining a situation where someone might want to skip using UTXOs older than x blocks.

ishaanam commented:
@harding
It could be useful in the scenario in which someone doesn't want to reveal the age of their wallet (Eg. specify `maxconf` as 2016 confirmations so that the recipient doesn't know that the wallet dates all the way back to 10 years ago).
